### PR TITLE
docs: refactor extension sidebar

### PR DESCRIPTION
**Pull request by jannyHou**
Signed-off-by: jannyHou <juehou@ca.ibm.com>

After https://github.com/strongloop/loopback.io/pull/1079 merged. The sidebar will be:

"Extending LoopBack 4" and "Building Applications" are expanded:
Screen Shot 2020-10-12 at 10 28 34 PM

Extension point and extensions is moved to concepts
Screen Shot 2020-10-12 at 10 28 53 PM

## Checklist

- [ ] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [ ] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: docs/site/sidebars/lb4_sidebar.yml [hunk 1]
```diff
@@ -23,629 +23,731 @@ children:
   url: Inside-LoopBack-Application.html
   output: 'web, pdf'
 
-- title: 'Tutorials'
-  url: Tutorials.html
+- title: 'Concepts Explanation'
```
Comment (raymondfeng) [inline docs/site/sidebars/lb4_sidebar.yml:26]: I would like to keep `Concepts`.

Comment (bajtos) [reply docs/site/sidebars/lb4_sidebar.yml:26]: +1 to keep concepts

**Comment by bajtos** [comment]
Personally, I feel we should put Tutorials & How-to guides higher and Concepts lower in the list. IMO, most people start with reading practical guides first, to see how they can solve a problem at their hands, and only later dive deeper to build a better understanding of the concepts involved.

1. Inside a LoopBack application
2. Building applications
3. Building extensions
4. Concepts

I re-read the discussion in https://github.com/strongloop/loopback-next/issues/5785, I am afraid we did not reach consensus on this aspect in that thread 😞 

I also find the extra level of nesting (Concepts >> Core ...) as too much, can we remove Core section and put its children right under Concepts?

**Comment by bajtos** [comment]
BTW I find the diff pretty much impossible to review, is there a way how to split the changes into smaller commits that would be easier to review? I don't know how much effort that would require, maybe it's better to preview your proposed version locally and review the new sidebar instead of the diff of changes? 🤷🏻

File: docs/site/sidebars/lb4_sidebar.yml (hunk 1 above)
Comment (jannyHou) [reply docs/site/sidebars/lb4_sidebar.yml:26]: I like "Concepts" too. 👏

**Comment by jannyHou** [comment]
> I feel we should put Tutorials & How-to guides higher and Concepts lower in the list.

@bajtos I see we can continue discussion here, it's just one item's position.

> I also find the extra level of nesting (Concepts >> Core ...) as too much, can we remove Core section and put its children right under Concepts?

I don't have a strong opinion on that. I can create 2 screenshots for compare.


> BTW I find the diff pretty much impossible to review, is there a way how to split the changes into smaller commits that would be easier to review?

I can split the commits. Most of the changes are moving items under "Extensions" to "Extending LoopBack/How-to guides", there is only one menu change besides that. Yeah you can review the screenshots if that's easier for you. I had them in the description.

**Comment by jannyHou** [comment]
@raymondfeng @bajtos thank you for the feedback, I split the commits, please review with hide whitespace.

The latest screenshot:

Screen Shot 2020-10-15 at 10 31 52 AM

**DISCUSSION:**

cc @strongloop/loopback-admins 

Where does the "Concepts" go? Quoting the opinions here:

Option 1: right after "Inside a LoopBack application"

https://github.com/strongloop/loopback-next/issues/5785#issuecomment-700795850
> As they are common to applications and extensions

Option 2: stay as is (right before "Best Practice")

https://github.com/strongloop/loopback-next/pull/6553#issuecomment-707767141

> Personally, I feel we should put Tutorials & How-to guides higher and Concepts lower in the list. IMO, most people start with reading practical guides first, to see how they can solve a problem at their hands, and only later dive deeper to build a better understanding of the concepts involved.

> Inside a LoopBack application
Building applications
Building extensions
Concepts

**Comment by raymondfeng** [comment]
@jannyHou Here is my take:

1. The title `BUILDING LOOPBACK` is confusing as I read it as building the framework itself. I still prefer to have `BUILDING LOOPBACK APPLICATIONS`

2. `EXTENDING LOOPBACK 4` -> `EXTENDING LOOPBACK FRAMEWORK`?

3. Only expanding `BUILDING LOOPBACK` and `EXTENDING LOOPBACK 4` does not seem to be very helpful as the 2nd level titles for these two sections are still to generic.

4. I prefer to have `CONCEPTS` come right after `INSIDE A LOOPBACK APPLICATION`. For comparison, https://docs.nestjs.com/ have `OVERVIEW` and `FUNDAMENTALS`.

**Comment by bajtos** [comment]
> The title `BUILDING LOOPBACK` is confusing as I read it as building the framework itself. I still prefer to have `BUILDING LOOPBACK APPLICATIONS`

+1 to include "applications" in the section title

> 2. `EXTENDING LOOPBACK 4` -> `EXTENDING LOOPBACK FRAMEWORK`?

+1

> 3. Only expanding `BUILDING LOOPBACK` and `EXTENDING LOOPBACK 4` does not seem to be very helpful as the 2nd level titles for these two sections are still to generic. Maybe we don't need to have `BUILDING LOOPBACK APPLICATIONS` as the wrapper and flatten its child items as top level ones.

IMO, we need to find the right balance - if we expand everything then the sidebar would not fit into the single screen.

Personally, I find the current proposal as the right balance between expanding too much or too little.

I find most top-level section as rather generic (Concepts, Best practices, Reference guides), so I don't see any problem with always-expanded 2nd level titles being generic in the same way 🤷🏻 

> 4. I prefer to have `CONCEPTS` come right after `INSIDE A LOOPBACK APPLICATION`. For comparison, https://docs.nestjs.com/ have `OVERVIEW` and `FUNDAMENTALS`.

Please note Nestjs is following a different documentation system than we do. IMO, if a user new to LoopBack go from "Inside a LB app" to "Concepts", they will be confused, because there will be a lot of new concepts they may not be familiar with.

---

Anyhow, I don't want us to spend more time arguing about details than necessary. I tried to explain my point of view, I'll leave it up to you @jannyHou to decide if and how to incorporate it into the proposal.

**Comment by jannyHou** [comment]
> he title BUILDING LOOPBACK is confusing as I read it as building the framework itself. I still prefer to have BUILDING LOOPBACK APPLICATIONS

My bad, it should be building applications. I updated to "Building LoopBack Applications"

> EXTENDING LOOPBACK 4 -> EXTENDING LOOPBACK FRAMEWORK?

👌 

> IMO, we need to find the right balance - if we expand everything then the sidebar would not fit into the single screen.
Personally, I find the current proposal as the right balance between expanding too much or too little.
I find most top-level section as rather generic (Concepts, Best practices, Reference guides), so I don't see any problem with always-expanded 2nd level titles being generic in the same way 🤷🏻

Agree with @bajtos on this. Listing all the tutorials and guides looks messy and confusing, see https://github.com/strongloop/loopback.io/pull/1079#issuecomment-706394053, especially some tutorials have long titles.

**Review by raymondfeng — APPROVED** [review]

**Comment by jannyHou** [comment]
The new sidebar expands like this:

Screen Shot 2020-10-15 at 12 47 53 PM

**Comment by jannyHou** [comment]
For the “Concepts”, I am also not sure which place is better.

“Concepts” has “Crafting LoopBack 4” in it, which implies it should be right after “Inside lb app”, otherwise I tend to show users tutorials+guides first(what lb can do), then explain concepts(how lb is designed to do them).

I will leave the concept refactor out of this PR.

**Comment by dhmlau** [comment]
Related to https://github.com/strongloop/loopback-next/issues/6478

**Pull request merged**